design/observerability-2.md

```diff
@@ -0,0 +1,196 @@
+| Health                 | Error rate   | Monitor error rates of individual Keycloak services                           | To be defined | Investigate further                                      |
+| Health (optional)      | Error rate   | Monitor error rate in logs                                                    | To be defined | If increases, check what might be causing it             |
+| Performance            | Latency      | Monitor the aggregated average response time for individual Keycloak Services | To be defined | If drastically increases, check what might be causing it |
+
+For all the SLOs mentioned the above, the Keycloak Operator needs to set proper Prometheus Alerts.
+
+## Implementation
+
+[This document](https://docs.google.com/spreadsheets/d/1YGn4wwUupgQsFTmlm-5VXgLYkigTdy_TEnDhqFxNXXo/edit?usp=sharing) contains exported metrics from Keycloak's `/metrics` endpoint.
+
+The missing metrics are:
+- HTTP 5xx error rates for OIDC, SAML, Account Console and Admin Console endpoints
+- Aggregated latency for OIDC, SAML, Account Console and Admin Console endpoints
+- External IdP responses (especially 5xx)
+
+The additional metrics will be implemented using [Microprofile Metrics specification](https://github.com/eclipse/microprofile-metrics/blob/master/spec/src/main/asciidoc/metrics_spec.adoc).
+
+#### Enabling Metrics in Keycloak and Keycloak.x
+
+The Microprofile Metrics (technically [Smallrye Metrics](https://github.com/smallrye/smallrye-metrics)) are already
+integrated into Wildfly. From Keycloak perspective, we need to obtain `MetricRegistries` class and invoke
+`MetricRegistries#get(MetricRegistry.Type.APPLICATION)` method to obtain a registry. Once we do that, we can
+obtain all instruments for adding additional metrics.
+
+The `MetricRegistries` object obtained from Wildfly needs to be the same as the one used by the
+`org.eclipse.microprofile.metrics.api` module. Therefore, we need to update the `module.xml` used by the
+`org.keycloak.keycloak-server-subsystem.dependencies`
+(see [here](https://github.com/keycloak/keycloak/blob/cee884e4a72c08b7ecafd965ab6f8c430b17ef15/distribution/feature-packs/server-feature-pack/src/main/resources/modules/system/layers/keycloak/org/keycloak/keycloak-server-subsystem/main/server-war/WEB-INF/jboss-deployment-structure.xml)).
+
+The Quarkus-based distribution (Keycloak.x) requires adding `smallrye-metrics` extension (see [the manual](https://quarkus.io/guides/microprofile-metrics)).
+Once this is done, the metrics will be automatically available in `/metrics` endpoint. From the implementation perspective,
+we need to perform the same steps as we were on Wildfly.
+
+## Metrics for the REST interface
+
+Most of the metrics are related to the various REST endpoints. Therefore, it makes sense to put the root metrics
+component into `services` module under its own package - `org.keycloak.services.metrics`.
```

Comment:
We may want to introduce our own abstraction on top of smallrye/microprofile metrics. Also, we probably want to have the ability to plug-in custom metrics providers into Keycloak.

Reply:
Further, we should have 1 health endpoint, and 1 metrics endpoint. This should be covered in this design.

Reply:
> We may want to introduce our own abstraction on top of smallrye/microprofile metrics. 

I don't like this. We have no idea if Micrometer will implement Microprofile Metrics or not. If it will, implementing our own metrics stuff on top of Microprofile Metrics (instead of calling Smallrye directly) would be enough. Let's keep it simple and follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) rule. 

> Also, we probably want to have the ability to plug-in custom metrics providers into Keycloak.

Could you please elaborate? What Metrics Providers do you have in mind? If you're thinking about Micrometer - it depends what the future will bring us. Either Micrometer will support Microprofile Metrics and there will be little-to-none modifications required on our side (we'll just swap the dependencies). Or we'll need adapt to Micrometer by rewriting our code. Unfortunately, it's very early stage of this discussions (they mainly happen within Quarkus Team) and we know nothing at this point. 

> Further, we should have 1 health endpoint, and 1 metrics endpoint. This should be covered in this design.

Yes, I'll put that into the design.

Reply:
I believe this one has been resolved. Please reopen if you think otherwise.